compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECreateReferenceToAllocation.cpp

```diff
@@ -0,0 +1,69 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-amd-aie/IR/AMDAIEDialect.h"
+#include "iree-amd-aie/IR/AMDAIEOps.h"
+#include "iree-amd-aie/Transforms/AMDAIEUtils.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "mlir/IR/Iterators.h"
+#include "mlir/Pass/Pass.h"
+
+#define DEBUG_TYPE "iree-amdaie-create-reference-to-allocation"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+class AMDAIECreateReferenceToAllocationPass
+    : public impl::AMDAIECreateReferenceToAllocationBase<
+          AMDAIECreateReferenceToAllocationPass> {
+ public:
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<AMDAIEDialect, memref::MemRefDialect>();
+  }
+
+  AMDAIECreateReferenceToAllocationPass() = default;
+  AMDAIECreateReferenceToAllocationPass(
+      const AMDAIECreateReferenceToAllocationPass &pass){};
+  void runOnOperation() override;
+};
+
+void AMDAIECreateReferenceToAllocationPass::runOnOperation() {
+  MLIRContext *context = &getContext();
+  mlir::FunctionOpInterface funcOp = getOperation();
+  IRRewriter rewriter(context);
+
+  funcOp->walk([&](memref::AllocOp allocOp) {
```

Comment:
What are the assumptions on the `memref.alloc` ops? We are assuming here that they are already completely hoisted to the global scope? I think we should make sure that the `reference_to` ops are inserted locally and while the `memref.alloc` ops should be global.

Related, I think we have some test cases with nested loops, even a peeled version, to make sure the output is in the shape we need it to be (references local, allocs global).

Reply:
> What are the assumptions on the memref.alloc ops? We are assuming here that they are already completely hoisted to the global scope? I think we should make sure that the reference_to ops are inserted locally and while the memref.alloc ops should be global.

Nvm, I see this should be inserted before `HoistStaticallyBoundAllocations`

Reply:
Exactly. Could you check a complete IR here?https://gist.github.com/yzhang93/25288d8299e75aeba47775d761dbadbe.

Reply:
Cool, maybe it's useful to add a test where we run both passes?